vector/src/main/java/im/vector/app/features/displayname/Extension.kt

```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright (c) 2021 New Vector Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package im.vector.app.features.displayname
+
+import org.matrix.android.sdk.api.util.MatrixItem
+
+fun MatrixItem.getBestName(): String {
```

Comment:
QUESTION: what's the difference between `MatrixItem.getBestName()` and `DisplayNameResolver.getBestName()`, do we need both?

Reply:
That's a good question, and I do not like the code duplication too.
The idea was to create MatrixItem.getBestName() extension on the app side, to avoid touching the code everywhere.
If you find a way to get `displayNameResolver` here, it would be great.

Reply:
I can think of two options but they both need more code changes, I'm not sure if they're worth it.

- extract all the `MatrixItem` creation to a dedicated factory, passing and creating the best name at the point of item creation

```kotlin
class MatrixItemFactory @Inject constructor() {
  fun createUser() = MatrixItem.User(bestName = "calculate best name")
  fun createGroup() = MatrixItem.Group(bestName = "calculate best name")
  //etc...
}
```

- expose the name resolver on the `Session` or a `Service` and use a `with` to access the extension

```kotlin
val bestName = with(session) {
  MatrixItem.getBestName()
}


interface Session/*Service {
  fun displayNameResolver()
  fun MatrixItem.getBestName() = displayNameResolver.getBestName(this)
}
```


I guess our only potential issue with the current implementation is the `matrixItemDisplayNameFallbackProvider` and `VectorMatrixItemDisplayNameFallbackProvider` becoming inconsistent

Reply:
Thanks for the suggestions.

> I guess our only potential issue with the current implementation is the matrixItemDisplayNameFallbackProvider and VectorMatrixItemDisplayNameFallbackProvider becoming inconsistent

I agree, maybe just add a comment in the Vector side would be enough.

Reply:
https://github.com/vector-im/element-android/pull/4034/commits/b5f5f8cc73b135e8385abb6bf9911d0759a7e7d0

Reply:
thanks!